--- qtdeclarative-opensource-src-5.9.0-alpha/src/3rdparty/masm/assembler/AbstractMacroAssembler.h.omv~	2017-03-15 20:21:23.585893913 +0100
+++ qtdeclarative-opensource-src-5.9.0-alpha/src/3rdparty/masm/assembler/AbstractMacroAssembler.h	2017-03-15 20:21:38.441951250 +0100
@@ -809,14 +809,14 @@ public:
     {
         AssemblerType::cacheFlush(code, size);
     }
+
+    AssemblerType m_assembler;
 protected:
     AbstractMacroAssembler()
         : m_randomSource(cryptographicallyRandomNumber())
     {
     }
 
-    AssemblerType m_assembler;
-    
     uint32_t random()
     {
         return m_randomSource.getUint32();
--- qtdeclarative-opensource-src-5.9.0-alpha/src/plugins/scenegraph/scenegraph.pro.omv~	2017-03-15 21:04:30.357707785 +0100
+++ qtdeclarative-opensource-src-5.9.0-alpha/src/plugins/scenegraph/scenegraph.pro	2017-03-15 21:04:42.800764541 +0100
@@ -1,5 +1,5 @@
 TEMPLATE = subdirs
 QT_FOR_CONFIG += quick
 qtConfig(d3d12): SUBDIRS += d3d12
-qtConfig(openvg): SUBDIRS += openvg
+#qtConfig(openvg): SUBDIRS += openvg
 
