diff -up qtdeclarative-everywhere-src-5.14.1/tools/qmlplugindump/main.cpp.omv~ qtdeclarative-everywhere-src-5.14.1/tools/qmlplugindump/main.cpp
--- qtdeclarative-everywhere-src-5.14.1/tools/qmlplugindump/main.cpp.omv~	2020-02-06 22:01:01.150485050 +0100
+++ qtdeclarative-everywhere-src-5.14.1/tools/qmlplugindump/main.cpp	2020-02-06 22:01:53.278487349 +0100
@@ -110,12 +110,12 @@ struct QmlVersionInfo
 
 static bool matchingImportUri(const QQmlType &ty, const QmlVersionInfo& versionInfo) {
     if (versionInfo.strict) {
-        return (versionInfo.pluginImportUri == ty.module()
+        return (QHashedString(versionInfo.pluginImportUri) == ty.module()
                 && (ty.majorVersion() == versionInfo.majorVersion || ty.majorVersion() == -1))
                 || ty.module().isEmpty();
     }
     return ty.module().isEmpty()
-            || versionInfo.pluginImportUri == ty.module()
+            || QHashedString(versionInfo.pluginImportUri) == ty.module()
             || ty.module().startsWith(versionInfo.pluginImportUri + QLatin1Char('.'));
 }
 
