diff --git a/src/qml/util/qqmladaptormodel.cpp b/src/qml/util/qqmladaptormodel.cpp
index 5fc2444..0c37918 100644
--- a/src/qml/util/qqmladaptormodel.cpp
+++ b/src/qml/util/qqmladaptormodel.cpp
@@ -171,8 +171,14 @@ public:
                 signalIndexes.append(propertyId + signalOffset);
         }
 
-        for (int i = 0, c = items.count();  i < c; ++i) {
-            QQmlDelegateModelItem *item = items.at(i);
+        const QList<QQmlDelegateModelItem *> copy = items;
+        for (int i = 0, c = copy.count();  i < c; ++i) {
+            // Applying the same logic used in QQmlDelegateModel::_q_itemsRemoved().
+            QQmlDelegateModelItem *item = copy.at(i);
+            if (!items.contains(item)) {
+                continue;
+            }
+
             const int idx = item->modelIndex();
             if (idx >= index && idx < index + count) {
                 for (int i = 0; i < signalIndexes.count(); ++i)
