diff -up qtdeclarative-everywhere-src-5.12.0/src/3rdparty/masm/masm.pri.2~ qtdeclarative-everywhere-src-5.12.0/src/3rdparty/masm/masm.pri
--- qtdeclarative-everywhere-src-5.12.0/src/3rdparty/masm/masm.pri.2~	2018-12-17 00:39:06.007622442 +0100
+++ qtdeclarative-everywhere-src-5.12.0/src/3rdparty/masm/masm.pri	2018-12-17 00:40:04.221240772 +0100
@@ -107,7 +107,7 @@ debug_and_release {
         GENERATEDDIR = $$GENERATEDDIR/release
     }
 }
-INCLUDEPATH += $$GENERATEDDIR
+INCLUDEPATH += $$OUT_PWD/$$GENERATEDDIR
 
 retgen.output = $$GENERATEDDIR/RegExpJitTables.h
 retgen.script = $$PWD/yarr/create_regex_tables
