diff -up qtdeclarative-everywhere-src-5.15.1/src/quick/scenegraph/coreapi/qsgshaderrewriter.cpp.omv~ qtdeclarative-everywhere-src-5.15.1/src/quick/scenegraph/coreapi/qsgshaderrewriter.cpp
--- qtdeclarative-everywhere-src-5.15.1/src/quick/scenegraph/coreapi/qsgshaderrewriter.cpp.omv~	2020-09-13 19:35:15.476826358 +0200
+++ qtdeclarative-everywhere-src-5.15.1/src/quick/scenegraph/coreapi/qsgshaderrewriter.cpp	2020-09-13 19:36:01.912376292 +0200
@@ -192,6 +192,10 @@ QByteArray qsgShaderRewriter_insertZAttr
     QByteArray result;
     result.reserve(1024);
     result += QByteArray::fromRawData(input, voidPos - input);
+#if defined(QT_OPENGL_ES_2)
+    result += "uniform highp float _qt_order;\n"
+              "uniform highp float _qt_zRange;\n";
+#else
     switch (profile) {
     case QSurfaceFormat::NoProfile:
     case QSurfaceFormat::CompatibilityProfile:
@@ -204,6 +208,7 @@ QByteArray qsgShaderRewriter_insertZAttr
                   "uniform float _qt_zRange;\n";
         break;
     }
+#endif
 
     // Find first brace '{'
     while (t != Tokenizer::Token_EOF && t != Tokenizer::Token_OpenBrace) t = tok.next();
